Enhancement: Featured Categories #6247

TMCSherpa opened the pull request:
Take 2
All the DB changes are now in mysql_zencart.sql so composer should run fine this time.
Names and dates have been rolled back so that shouldn't be a problem
A list of DB changes are in the updates folder so that file can get generated.

lat9 commented on includes/templates/template_default/templates/tpl_index_default.php:
```diff
@@ -68,6 +72,15 @@
+ * display the Featured Products Center Box
+ */
+?>
+<?php require($template->get_template_dir('tpl_modules_featured_categories.php',DIR_WS_TEMPLATE, $current_page_base,'templates'). '/tpl_modules_featured_categories.php'); ?>
```
- Remove leading/trailing parentheses around required file.
- Add single-space between function arguments and after function (i.e. before the concatenating `.`).

lat9 commented on includes/templates/template_default/templates/tpl_modules_featured_categories.php:
```diff
@@ -0,0 +1,26 @@
+ * based on tpl_modules_featured_products
+ */
+  $zc_show_featured = false;
+  include(DIR_WS_MODULES . zen_get_module_directory(FILENAME_FEATURED_CATEGORIES_MODULE));
```
Remove leading/trailing parentheses.

lat9 commented on includes/templates/template_default/templates/tpl_modules_featured_categories.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * require the list_box_content template to display the category
+ */
+  require($template->get_template_dir('tpl_columnar_display.php',DIR_WS_TEMPLATE, $current_page_base,'common'). '/tpl_columnar_display.php');
```
- Remove leading/trailing parentheses around required file.
- Add single-space between function arguments and after function (i.e. before the concatenating `.`).

lat9 replied:
Still needs the single space between function arguments:

```php
  require $template->get_template_dir('tpl_columnar_display.php', DIR_WS_TEMPLATE, $current_page_base, 'common') . '/tpl_columnar_display.php';
```

lat9 commented on zc_install/sql/demo/demo_cleanup.sql:
```diff
@@ -37,4 +37,5 @@ TRUNCATE TABLE record_company_info;
 TRUNCATE TABLE salemaker_sales; 
-TRUNCATE TABLE specials; 
+TRUNCATE TABLE specials;
+TRUNCATE TABLE featured_categories; 
```
Each of the table truncations in this file are alphabetically ordered based on the table's name.

lat9 commented on zc_install/sql/install/mysql_zencart.sql:
```diff
@@ -845,6 +845,29 @@ CREATE TABLE featured (
+DROP TABLE IF EXISTS featured_categories;
+CREATE TABLE featured_categories (
+  featured_categories_id int(11) NOT NULL auto_increment,
+  categories_id int(11) NOT NULL default '0',
```
The value's an `int`, its default should also be an int ... i.e. `0` instead of `'0'`.

lat9 replied:
Oops, the `status` field should be similarly changed to `default 1` instead of `default '1'`.

lat9 commented on zc_install/sql/install/zero_dates_cleanup.sql:
```diff
@@ -94,3 +94,8 @@ UPDATE record_artists_info SET date_last_click = NULL WHERE CAST(date_last_click
 UPDATE record_company SET date_added = NULL WHERE CAST(date_added AS CHAR(19)) = '0000-00-00 00:00:00';
 UPDATE record_company SET last_modified = NULL WHERE CAST(last_modified AS CHAR(19)) = '0000-00-00 00:00:00';
 UPDATE record_company_info SET date_last_click = NULL WHERE CAST(date_last_click AS CHAR(19)) = '0000-00-00 00:00:00';
+UPDATE featured_categories SET featured_date_added = NULL WHERE CAST(featured_date_added AS CHAR(19)) = '0000-00-00 00:00:00';
+UPDATE featured_categories SET featured_last_modified = NULL WHERE CAST(featured_last_modified AS CHAR(19)) = '0000-00-00 00:00:00';
+UPDATE featured_categories SET expires_date = '0001-01-01' WHERE CAST(expires_date AS CHAR(10)) = '0000-00-00';
+UPDATE featured_categories SET date_status_change = NULL WHERE CAST(date_status_change AS CHAR(19)) = '0000-00-00 00:00:00';
+UPDATE featured_categories SET featured_date_available = '0001-01-01' WHERE CAST(featured_date_available AS CHAR(10)) = '0000-00-00';
```
Not needed, since that table's newly added.

lat9 commented on zc_install/sql/install/mysql_zencart.sql:
```diff
@@ -845,6 +845,29 @@ CREATE TABLE featured (
+  featured_date_available date NOT NULL default '0001-01-01',
+  PRIMARY KEY  (featured_categories_id),
+  KEY idx_status_zen (status),
+  KEY idx_products_id_zen (categories_id),
```
Shouldn't this be `idx_categories_id_zen`?

lat9 commented on zc_install/sql/install/mysql_zencart.sql:
```diff
@@ -3131,6 +3155,9 @@ INSERT INTO configuration (configuration_title, configuration_key, configuration
 INSERT INTO configuration (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, last_modified, date_added, use_function, set_function) VALUES ('EZ-Pages Pages to disable footers', 'EZPAGES_DISABLE_FOOTER_DISPLAY_LIST', '', 'EZ-Pages "pages" on which to NOT display the normal "footer" for your site.<br />Simply list page ID num ...
 INSERT INTO configuration (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, last_modified, date_added, use_function, set_function) VALUES ('EZ-Pages Pages to disable left-column', 'EZPAGES_DISABLE_LEFTCOLUMN_DISPLAY_LIST', '', 'EZ-Pages "pages" on which to NOT display the normal "left" column (of sideboxes) for your site.< ...
 INSERT INTO configuration (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, last_modified, date_added, use_function, set_function) VALUES ('EZ-Pages Pages to disable right-column', 'EZPAGES_DISABLE_RIGHTCOLUMN_DISPLAY_LIST', '', 'EZ-Pages "pages" on which to NOT display the normal "right" column (of sideboxes) for your sit ...
+INSERT INTO configuration (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, last_modified, date_added, use_function, set_function, val_function) VALUES ('Show Featured Categories on Main Page', 'SHOW_PRODUCT_INFO_MAIN_FEATURED_CATEGORIES', '5', 'Show Featured Categories on Main Page<br />0= off or set the sort order', 24,  ...
+INSERT INTO configuration (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, last_modified, date_added, use_function, set_function, val_function) VALUES ('Random Featured Categories for SideBox', 'MAX_RANDOM_SELECT_FEATURED_CATEGORIES', '1', 'Number of random FEATURED categories to rotate in the sidebox<br />Enter the numbe ...
```
Use HTML5 `<br>` instead of `<br />`.

lat9 commented on zc_install/sql/install/mysql_zencart.sql:
```diff
@@ -2399,7 +2422,8 @@ VALUES ('configMyStore', 'BOX_CONFIGURATION_MY_STORE', 'FILENAME_CONFIGURATION',
        ('mediaManager', 'BOX_CATALOG_MEDIA_MANAGER', 'FILENAME_MEDIA_MANAGER', '', 'extras', 'Y', 4),
-       ('mediaTypes', 'BOX_CATALOG_MEDIA_TYPES', 'FILENAME_MEDIA_TYPES', '', 'extras', 'Y', 5);
+       ('mediaTypes', 'BOX_CATALOG_MEDIA_TYPES', 'FILENAME_MEDIA_TYPES', '', 'extras', 'Y', 5),
+       ('featured_categories', 'BOX_CATALOG_FEATURED_CATEGORIES', 'FILENAME_FEATURED_CATEGORIES', '', 'catalog', 'Y', 13);
```
This section groups the inserted values based on the `menu_key`; move this up with the rest of the `'catalog'` items.

lat9 commented on includes/templates/template_default/templates/tpl_featured_categories_default.php:
```diff
@@ -0,0 +1,45 @@
+        $col = 0;
+        foreach ($listing as $record) {
+            $lc_text = '<a href="' . zen_href_link(FILENAME_DEFAULT, 'categories_id=' . $record['categories_id']) . '">';
+            $lc_text .= '<a href="' . zen_href_link(FILENAME_DEFAULT, 'cPath='.  zen_get_generated_category_path_rev($record["categories_id"])). '">'
```
Use single-, not double-, quotes around the `categories_id`.

lat9 commented on includes/templates/template_default/sideboxes/tpl_featured_categories.php:
```diff
@@ -0,0 +1,25 @@
+    $data = (new Category((int)$random_featured_categories->fields['categories_id']))->withDefaultLanguage()->getData();
+    $featured_category_box_counter++;
+    $content .= "\n" . '  <div class="sideBoxContentItem">';
+    $content .= '<a href="' . zen_href_link(FILENAME_DEFAULT, 'cPath='.  zen_get_generated_category_path_rev($data["categories_id"])). '">'
```
- Use single-, not double-, quotes around the `categories_id`.
- Add single space before the `.` concatenation operator; remove extra space after.

lat9 replied:
Still missing the single-space before the various concatenation operators.

lat9 commented on includes/templates/responsive_classic/templates/tpl_modules_mobile_menu.php:
```diff
@@ -88,6 +88,15 @@
     }
   }
 ?>
+<?php if (SHOW_CATEGORIES_BOX_FEATURED_PRODUCTS == 'true') {
```
Are the two "SHOW_CATEGORIES_BOX_*" definitions reversed?

lat9 commented on includes/templates/responsive_classic/sideboxes/tpl_categories.php:
```diff
@@ -86,6 +86,12 @@
         $content .= '<li><a class="category-links" href="' . zen_href_link(FILENAME_FEATURED_PRODUCTS) . '">' . CATEGORIES_BOX_HEADING_FEATURED_PRODUCTS . '</a></li>' . "\n";
       }
     }
+    if (SHOW_CATEGORIES_BOX_FEATURED_CATEGORIES== 'true') {
```
- Needs a space before the comparison operator.
- Use `===`.

lat9 replied:
Changes still needed.

lat9 commented on includes/modules/sideboxes/featured_categories.php:
```diff
@@ -0,0 +1,36 @@
+    if ($random_featured_categories->RecordCount() > 0)  {
+      require($template->get_template_dir('tpl_featured_categories.php',DIR_WS_TEMPLATE, $current_page_base,'sideboxes'). '/tpl_featured_categories.php');
+      $title =  BOX_HEADING_FEATURED_CATEGORIES;
+      require($template->get_template_dir($column_box_default, DIR_WS_TEMPLATE, $current_page_base,'common') . '/' . $column_box_default);
```
- Remove leading/trailing parentheses around required file.
- Add single-space between function arguments and after function (i.e. before the concatenating `.`).

lat9 replied:
Still need spaces after each comma in the function arguments' list.

lat9 commented on includes/modules/pages/featured_categories/header_php.php:
```diff
@@ -0,0 +1,54 @@
+// These are "soft" configuration setting that can be overridden on a site-specific basis.
+//
+if (!defined('FC_MAX_COLUMNS')) {
+    define('FC_MAX_COLUMNS', '6');
```
Why not create a setting in the `configuration` table, like is done for the maximum sidebox entries to display?

lat9 commented on admin/includes/languages/lang.english.php:
```diff
@@ -679,5 +679,8 @@
     '_OCTOBER' => 'October',
     '_SEPTEMBER' => 'September',
+    'BOX_CATALOG_FEATURED_CATEGORIES' => 'Featured Categories',
+    'IMAGE_EDIT_CATEGORY' => 'Edit Category',
```
Array elements in this file are organized alphabetically by 'key'.

lat9 reviewed (changes requested):
Great start, but I'm _very_ nit-picky!

TMCSherpa commented:
I've made most of the changes locally and added a line to functions_categories.php so the database gets cleaned up if a category is removed. 
For some reason my boss expects me to do some work so I'll finish it up tonight, test it and upload the changes to the branch

TMCSherpa commented:
Even more alphabetical. Added css tags, removed all references to products other than SHOW_PRODUCT_INFO_COLUMNS_FEATURED_CATEGORIES because RLIKE breaks. I think? Maybe it's only SHOW_PRODUCT_INFO_MAIN_FEATURED_CATEGORIES that needs it. It was a big enough pain to figure that out I'm not going to check

lat9 commented on includes/modules/pages/featured_categories/header_php.php:
```diff
@@ -0,0 +1,57 @@
+
+
+// This should be last line of the script:
+$zco_notifier->notify('NOTIFY_HEADER_START_FEATURED_CATEGORIES', null);
```
This should be `NOTIFY_HEADER_END_FEATURED_CATEGORIES`.

lat9 commented on includes/classes/Category.php:
```diff
@@ -0,0 +1,168 @@
+        $this->initLanguages();
+
+        if ($this->categories_id !== null) {
+            $this->data = $this->loadProductDetails($this->categories_id);
```
Shouldn't this method be named `loadCategoryDetails`?  There's nothing product-specific in the data returned.

lat9 commented on includes/classes/Category.php:
```diff
@@ -0,0 +1,168 @@
+
+        if ($product->EOF) {
+            return [];
+        }
```
- Since it's a single table in the query, you don't need the alias
- Use `$category` as the returned variable name; it's a category after all, not a product.

lat9 commented on includes/classes/Category.php:
```diff
@@ -0,0 +1,168 @@
+         */
+        $sql = "SELECT pd.*
+                FROM " . TABLE_CATEGORIES_DESCRIPTION . " pd
+                WHERE pd.categories_id = " . (int)$categories_id . "
```
- Since it's a single table in the query, you don't need the alias

lat9 commented on includes/classes/Category.php:
```diff
@@ -0,0 +1,168 @@
+        $data = $product->fields;
+        $data['id'] = $data['categories_id'];
+        $data['categories_id'] = $data['categories_id'];
+        //$data['parent_category_id'] = $data['master_categories_id'];
```
Remove comment; that field isn't associated with a category definition anyways.

lat9 commented on includes/modules/pages/featured_categories/header_php.php:
```diff
@@ -0,0 +1,57 @@
+                FROM " . TABLE_CATEGORIES . " p
+                LEFT JOIN " . TABLE_FEATURED_CATEGORIES . " f ON p.categories_id = f.categories_id
+                LEFT JOIN " . TABLE_CATEGORIES_DESCRIPTION . " pd ON p.categories_id = pd.categories_id
+                AND pd.language_id = " . (int)$_SESSION['languages_id'] . "
```
Use `c` and `cd` as the aliases instead of `p` and `pd`; those are references to category information, not product.

lat9 commented on includes/modules/pages/featured_categories/header_php.php:
```diff
@@ -0,0 +1,57 @@
+// These are "soft" configuration setting that can be overridden on a site-specific basis.
+//
+if (!defined('SHOW_PRODUCT_INFO_COLUMNS_FEATURED_CATEGORIES')) {
+    define('SHOW_PRODUCT_INFO_COLUMNS_FEATURED_CATEGORIES', '3');
```
Use `zen_define_default`

lat9 replied:
Right, this is a database setting now (my bad) ... but it's not added in `zc_install/sql/updates/mysql_upgrade_zencart_210.sql`.

lat9 commented on includes/modules/sideboxes/featured_categories.php:
```diff
@@ -0,0 +1,37 @@
+                           left join " . TABLE_FEATURED_CATEGORIES . " f on p.categories_id = f.categories_id
+                           left join " . TABLE_CATEGORIES_DESCRIPTION . " pd on p.categories_id = pd.categories_id )
+                           where p.categories_id = f.categories_id
+                           and p.categories_id = pd.categories_id
```
- Use `c` and `cd` as the aliases instead of `p` and `pd`.
- Uppercase SQL keywords, e.g. `select` -> 'SELECT'
- No quotes around the `language_id` clause value, it's an integer.

lat9 replied:
Still need to remove quotes around the `language_id` clause value.

lat9 commented on includes/modules/sideboxes/featured_categories.php:
```diff
@@ -0,0 +1,37 @@
+ * @version $Id: DrByte 2020 Jul 10 Modified in v1.5.8-alpha $
+ * based on featured and brands
+ */
+
```
Overall, adhere to PSR12 indentation standards.  Code initially starts with no indentation, then 4 spaces for conditional clauses.

lat9 commented on includes/templates/template_default/sideboxes/tpl_featured_categories.php:
```diff
@@ -0,0 +1,25 @@
+    $content .= '</div>';
+    $random_featured_categories->MoveNextRandom();
+}
+$content .= '</div>' . "\n";
```
Empty line needed at the end-of-file

lat9 replied:
Close enough; now there are two.

lat9 commented on includes/templates/template_default/templates/tpl_featured_categories_default.php:
```diff
@@ -0,0 +1,45 @@
+
+            $col++;
+            if ($col >= SHOW_PRODUCT_INFO_COLUMNS_FEATURED_CATEGORIES) {
+                $col = 0;
```
Remove unwanted indentation.

lat9 commented on includes/templates/template_default/templates/tpl_modules_featured_categories.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * require the list_box_content template to display the category
+ */
+  require $template->get_template_dir('tpl_columnar_display.php',DIR_WS_TEMPLATE, $current_page_base,'common') . '/tpl_columnar_display.php';
```
Function parameters should be separated by `, `.

lat9 commented on includes/templates/template_default/templates/tpl_modules_featured_categories.php:
```diff
@@ -0,0 +1,26 @@
+?>
+</div>
+<?php } ?>
+<!-- eof: featured categories  -->
```
Add empty line at end-of file

lat9 commented on includes/classes/Category.php:
```diff
@@ -0,0 +1,168 @@
+
+        $this->languages = $lng->get_language_list();  // [1 => 'en', 2 => 'fr']
+    }
+}
```
All .php files should end with an empty line.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+$action = (isset($_GET['action']) ? $_GET['action'] : '');
+$currentPage = (isset($_GET['page']) && $_GET['page'] != '' ? (int)$_GET['page'] : 0);
+
+if (!empty($action)) {
```
Adhere to PSR12 standards, 4-space indentation for conditionals.

lat9 replied:
Use `if ($action !== '') {`

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+ */
+require 'includes/application_top.php';
+
+$action = (isset($_GET['action']) ? $_GET['action'] : '');
```
Use `$action = $_GET['action'] ?? '';

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+                    SET featured_last_modified = now(),
+                        expires_date = '" . zen_db_input($expires_date) . "',
+                        featured_date_available = '" . zen_db_input($featured_date_available) . "'
+                    WHERE featured_categories_id = " . (int)$featured_categories_id);
+
+        zen_redirect(zen_href_link(FILENAME_FEATURED_CATEGORIES, ($currentPage != 0 ? 'page=' . $currentPage . '&' : '') . (isset($_GET['search']) ? 'search=' . $_GET['search'] . '&' : '') . 'fID=' . (int)$featured_categories_id));
```
`(int)` casting of `$featured_categories_id` not needed here; that was already done at line 92.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+      $featured_categories_id = (int)$_POST['fID'];
+
+      $db->Execute("DELETE FROM " . TABLE_FEATURED_CATEGORIES . "
+                    WHERE featured_categories_id = " . (int)$featured_categories_id);
```
Don't need the cast here, it was just performed in the line above.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+                FROM " . TABLE_CATEGORIES . "
+                WHERE categories_id = " . (int)$_POST['pre_add_categories_id'];
+        $check_featured = $db->Execute($sql);
+        if ($check_featured->RecordCount() < 1) {// check for valid CID
```
Use `if ($check_featured->EOF) {`

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+          $featureds = $db->Execute("SELECT p.categories_id
+                                     FROM " . TABLE_CATEGORIES . " p,
+                                          " . TABLE_FEATURED_CATEGORIES . " f
+                                     WHERE f.categories_id = p.categories_id");
```
Use `c` and `cd` as aliases, not `p` and `pd`, these are categories tables, not products.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,550 @@
+  </body>
+</html>
+<?php
+require DIR_WS_INCLUDES . 'application_bottom.php';
```
Add empty line at end-of-file

lat9 commented on admin/includes/languages/english/lang.featured_categories.php:
```diff
@@ -0,0 +1,34 @@
+    'ERROR_INVALID_EXPIRES_DATE' => 'The &quot;Expires&quot; date is not valid, please re-enter.',
+];
+
+return $define;
```
Add empty line at end-of-file

TMCSherpa commented:
someone else can fix admin/featured.php because that's not in my scope.
I think I got everything? testing is working so I think all of the p. to c. and f. to fc. have been handled

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                                ];
+                                $contents[] = [
+                                'align' => 'text-center',
+                                'text' => '<a href="' . zen_href_link(FILENAME_CATEGORIES, '&cPath=&cID=' .$fInfo->categories_id . '&action=edit_category' ). '" class="btn btn-primary" role="button">' . IMAGE_EDIT_CATEGORY . '</a>'
```
- Why the empty `&cPath` parameter?
- Add spaces around all concatenation operators.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                    switch ($action) {
+                        case 'delete':
+                            $heading[] = ['text' => '<h4>' . TEXT_INFO_HEADING_DELETE_FEATURED . '</h4>'];
+                            $contents = ['form' => zen_draw_form('featured', FILENAME_FEATURED_CATEGORIES, 'action=deleteconfirm' . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : '')) . zen_draw_hidden_field('fID', $fInfo->featured_categories_id)];
```
Use `$page_search_parameters`.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                            $contents = ['form' => zen_draw_form('featured', FILENAME_FEATURED_CATEGORIES, 'action=deleteconfirm' . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : '')) . zen_draw_hidden_field('fID', $fInfo->featured_categories_id)];
+                            $contents[] = ['text' => TEXT_INFO_DELETE_INTRO];
+                            $contents[] = ['text' => '<b>' . zen_clean_html($fInfo->categories_name) . '"</b>'];
+                            $contents[] = ['align' => 'text-center', 'text' => '<br><button type="submit" class="btn btn-danger">' . IMAGE_DELETE . '</button> <a href="' . zen_href_link(FILENAME_FEATURED_CATEGORIES, 'fID=' . $fInfo->featured_categories_id . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : '')) . '" class="btn btn-defaul ...
```
Use `$page_search_parameters`.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+
+                        case 'pre_add':
+                            $heading[] = ['text' => '<h4>' . TEXT_INFO_HEADING_PRE_ADD_FEATURED . '</h4>'];
+                            $contents = ['form' => zen_draw_form('featured', FILENAME_FEATURED_CATEGORIES, 'action=pre_add_confirmation' . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : ''), 'post', 'class="form-horizontal"')];
```
Use `$page_search_parameters`.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                            $result = $db->Execute("SELECT MAX(categories_id) AS lastcategoryid FROM " . TABLE_CATEGORIES);
+                            $max_category_id = $result->fields['lastcategoryid'];
+                            $contents[] = ['text' => zen_draw_label(TEXT_ADD_FEATURED_SELECT, 'pre_add_categories_id', 'class="control-label"') . zen_draw_input_field('pre_add_categories_id', '', zen_set_field_length(TABLE_FEATURED_CATEGORIES, 'categories_id') . ' class="form-control" id="pre_add_categories_id" required max="' . $max_category_id . '"', '', 'number')];
+                            $contents[] = ['align' => 'text-center', 'text' => '<button type="submit" class="btn btn-primary">' . IMAGE_CONFIRM . '</button> <a href="' . zen_href_link(FILENAME_FEATURED_CATEGORIES, (!empty($fInfo->featured_categories_id) ? '&fID=' . $fInfo->featured_categories_id : '') . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_ ...
```
Use `$page_search_parameters`.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                                $contents[] = [
+                                'align' => 'text-center',
+                                'text' => '
+                                <a href="' . zen_href_link(FILENAME_FEATURED_CATEGORIES, '&fID=' . $fInfo->featured_categories_id . '&action=edit' . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : '')) . '" class="btn btn-primary" role="button">' . IMAGE_EDIT . '</a>
+                                <a href="' . zen_href_link(FILENAME_FEATURED_CATEGORIES, '&fID=' . $fInfo->featured_categories_id . '&action=delete' . ($currentPage != 0 ? '&page=' . $currentPage : '') . (isset($_GET['search']) ? '&search=' . $_GET['search'] : '')) . '" class="btn btn-warning" role="button">' . TEXT_INFO_HEADING_DELETE_FEATURED . '</a>'
```
Use `$page_search_parameters`.

TMCSherpa commented on includes/modules/sideboxes/featured_categories.php:
```diff
@@ -0,0 +1,38 @@
+        require $template->get_template_dir($column_box_default, DIR_WS_TEMPLATE, $current_page_base,'common') . '/' . $column_box_default;
+    }
+}
+?>
```
The code is from featured, it does the same thing

    // randomly select ONE featured product from the list retrieved:
    //$random_featured_product = zen_random_select($random_featured_products_query);
    $random_featured_product = $db->ExecuteRandomMulti($random_featured_products_query, MAX_RANDOM_SELECT_FEATURED_PRODUCTS);

    if ($random_featured_product->RecordCount() > 0)  {
      require($template->get_template_dir('tpl_featured.php',DIR_WS_TEMPLATE, $current_page_base,'sideboxes'). '/tpl_featured.php');
      $title =  BOX_HEADING_FEATURED_PRODUCTS;
      $title_link = FILENAME_FEATURED_PRODUCTS;
      require($template->get_template_dir($column_box_default, DIR_WS_TEMPLATE, $current_page_base,'common') . '/' . $column_box_default);
    }
  }
MAX_RANDOM_SELECT_FEATURED_PRODUCTS=2 on a fresh ZC install. For some reason

lat9 replied:
Yea, the whole `featured` element hasn't been loved for a while.  That ending `?>` needs to be replaced by a single empty line.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                          WHERE featured_categories_id = " . $featured_categories_id);
+
+            zen_redirect(zen_href_link(FILENAME_FEATURED_CATEGORIES, ($currentPage != 0 ? 'page=' . $currentPage . '&' : '') . (isset($_GET['search']) ? 'search=' . $_GET['search'] . '&' : '') . 'fID=' . $featured_categories_id));
+    break;
```
Still wrong indentation.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+
+    switch ($action) {
+
+    case 'setflag':
+        if (isset($_POST['flag']) && ($_POST['flag'] === '1' || $_POST['flag'] === '0')) {
+            zen_set_featured_status((int)$_POST['id'], (int)$_POST['flag']);
+            zen_redirect(zen_href_link(FILENAME_FEATURED_CATEGORIES, zen_get_all_get_params(['action', 'fID']) . 'fID=' . $_POST['id'], 'NONSSL'));
+        }
+        break;
```
Still not 4-space indented.

lat9 commented on admin/featured_categories.php:
```diff
@@ -0,0 +1,554 @@
+                                    if ((int)$item['featured_categories_id'] === (int)$_GET['fID']) {
+                                        break;
+                                    }
+                                $check_count++;
```
4-space indentation still needed.

lat9 commented:
If you'll just take care of those last few formatting change-requests, I'll actually install the code and put it through its paces.

TMCSherpa commented:
mysql_upgrade_zencart_210.sql didn't exist at the time so everything was in a file.
The upgrade changes have been added
I think I've made all the other changes but I had to add an update between starting to write this and hitting comment so who knows

lat9 commented on includes/templates/responsive_classic/sideboxes/tpl_categories.php:
```diff
@@ -8,93 +8,93 @@
-      }
+    if (SHOW_CATEGORIES_BOX_FEATURED_PRODUCTS === 'true') {
+        $show_this = $db->Execute("SELECT products_id FROM " . TABLE_FEATURED . " WHERE status= 1 limit 1");
+        if ($show_this->REOF) {
```
`->EOF`, not `->REOF`.

TMCSherpa replied:
Dang it

lat9 reviewed (changes requested):
Nice, just that one little typo.

lat9 reviewed (approved):
Close enough on those formatting changes.  Approving structurally, now it's time for functional validation!

TMCSherpa commented:
I found a couple things that probably don't matter but hey ho there you go

MAX_DISPLAY_SEARCH_RESULTS_FEATURED should probably be MAX_DISPLAY_SEARCH_RESULTS_FEATURED_CATEGORIES and have a new DB entry
same for MAX_DISPLAY_SEARCH_RESULTS_FEATURED_ADMIN

Product is depreciated so Category should also be depreciated even though I just added it?
It hasn't been changed in featured or specials yet so I can't copy that but it looks like it's just doing a complicated language_id check.
$data = $category_info->getDataForLanguage();

The boss isn't here so I can kind of work on it pending customers but it'll probably be tonight

I really want to add a module_ page_ sidebox_ prefix 'cus having the same file name in multiple locations is annoying but I'll stick to convention

drbyte commented:
> MAX_DISPLAY_SEARCH_RESULTS_FEATURED should probably be MAX_DISPLAY_SEARCH_RESULTS_FEATURED_CATEGORIES and have a new DB entry
same for MAX_DISPLAY_SEARCH_RESULTS_FEATURED_ADMIN

How likely is it that an admin is gonna want to change the number-per-page of displayed Featured Categories to be different than the number of Featured Products? 

(My thinking on this, and I know that some people think I hate admin-switches because of this stance, but here it is: No need for another switch if it'll never be changed. There's already too many per-page count switches IMO ... I've never seen a storeowner change any of them to be significantly different than others: In most cases they just wanted to set "all" of those to 100 or something large. I prefer to reduce the "clutter" of how many admin configurations are offered, because people get paralyzed by seeing too many and not knowing where to start or what does what. Again, that's just the rational "I" use when deciding whether to add another admin switch vs inherit from something already related and reasonable. Sometimes it's warranted, sometimes its debatable. FWIW.)

drbyte commented:
> Product is depreciated so Category should also be depreciated even though I just added it?

Can you explain further?

drbyte commented:
> I really want to add a module_ page_ sidebox_ prefix 'cus having the same file name in multiple locations is annoying but I'll stick to convention

Can you explain further?

lat9 commented:
> > I really want to add a module_ page_ sidebox_ prefix 'cus having the same file name in multiple locations is annoying but I'll stick to convention
> 
> Can you explain further?
> 
> Edit: I think I see what you mean. "featured.php" module was for products, and now you've added "featured_categories.php", which conflicts with the sidebox name. Maybe call the sidebox "categories_featured" (like there's a "categories" sidebox already, so use that prefix for context)?

In what way does that sidebox name-conflict arise?

drbyte commented:
> In what way does that sidebox name-conflict arise?

Just to whoever is looking at the filesystem.

/includes/modules/pages/featured_categories (page name)
/includes/modules/featured_categories (centerbox module)
/includes/modules/sideboxes/featured_categories (entirely different from the centerbox "module")

While the page name and centerbox module "conflict", that's minor because filenames don't compete.
But the centerbox and sidebox filenames "do" compete, and their content is vastly unrelated.

For Products, we have:
/includes/modules/featured_products
/includes/modules/sideboxes/featured
... so there's no apparent clash

I'd be okay with renaming /includes/modules/sideboxes/featured_categories.php to categories_featured.php, since it's akin to categories.php

TMCSherpa commented:
There is no naming conflict with featured(products). There are four files called featured_categories which would have been easier to reference in a pull request if they were
modules_featured_categories
page_featured_categories
sideboxes_featured_categories
It's fine, it was a joke.

TMCSherpa commented:
> > Product is depreciated so Category should also be depreciated even though I just added it?
> 
> Can you explain further?

psr4Autoload.php
    $psr4Autoloader->setClassFile('products', DIR_FS_CATALOG . DIR_WS_CLASSES . 'products.php');    //- Deprecated v2.1.0
classes/Categories is (more or less) the same as products

TMCSherpa commented:
> > MAX_DISPLAY_SEARCH_RESULTS_FEATURED should probably be MAX_DISPLAY_SEARCH_RESULTS_FEATURED_CATEGORIES and have a new DB entry
> > same for MAX_DISPLAY_SEARCH_RESULTS_FEATURED_ADMIN
> 
> How likely is it that an admin is gonna want to change the number-per-page of displayed Featured Categories to be different than the number of Featured Products?
> 
> (My thinking on this, and I know that some people think I hate admin-switches because of this stance, but here it is: No need for another switch if it'll never be changed. There's already too many per-page count switches IMO ... I've never seen a storeowner change any of them to be significantly different than others: In most cases they just wanted to set "all" of those to 100 or something large. I prefer to reduce the "clutter" of how many admin configurations are offered, because people get paralyzed by seeing too many and not knowing where to start or what does what. Again, that's just the rational "I" use when deciding whether to add another admin switch vs inherit from something already related and reasonable. Sometimes it's warranted, sometimes its debatable. FWIW.)

I added a few things copying settings for featured_products
SHOW_PRODUCT_INFO_COLUMNS_FEATURED_CATEGORIES
IMAGE_FEATURED_CATEGORY_LISTING_WIDTH
IMAGE_FEATURED_CATEGORY_LISTING_HEIGHT
CATEGORY_IMAGE_NO_IMAGE_STATUS
CATEGORY_IMAGE_NO_IMAGE

I can put them back if you want

drbyte commented:
> $psr4Autoloader->setClassFile('products', DIR_FS_CATALOG . DIR_WS_CLASSES . 'products.php'); //- Deprecated v2.1.0
classes/Categories is (more or less) the same as products

You might be mixing up `/includes/classes/products.php` (which is deprecated) 
vs `/includes/classes/Product.php` (which replaces `products.php`).
I think you've mostly cloned the new `Product.php` class, right? In that case, no deprecation issue.

drbyte commented:
> IMAGE_FEATURED_CATEGORY_LISTING_WIDTH
IMAGE_FEATURED_CATEGORY_LISTING_HEIGHT

I follow the logic of cloning that. I guess the question is: will someone sensible want to use a different-sized category image vs a product image? (In the main (non-"featured") product-listing "flow", where categories are presented for products to be selected from within those categories, `IMAGE_PRODUCT_LISTING_WIDTH` and `IMAGE_PRODUCT_LISTING_HEIGHT` are used for both products and categories.
Would it be suitable to simply update the "description" for `IMAGE_FEATURED_PRODUCTS_LISTING_WIDTH` and `IMAGE_FEATURED_PRODUCTS_LISTING_HEIGHT` to `Image - Featured Products/Categories Width` (and height, respectively)? Or is there a reasonable need for another pair of image-size settings?

drbyte commented:
> CATEGORY_IMAGE_NO_IMAGE_STATUS
CATEGORY_IMAGE_NO_IMAGE

I suppose these could be expanded to be used not only here with Featured Categories, but also in all the places where category images are used. Granted, it could be complicated to consistently determine whether it is a category vs a product image being affected, since the module that does it is kinda generic. 
It would need a bit of a code study ... which my brain isn't ready to look into tonight yet.

TMCSherpa commented:
> > $psr4Autoloader->setClassFile('products', DIR_FS_CATALOG . DIR_WS_CLASSES . 'products.php'); //- Deprecated v2.1.0
> > classes/Categories is (more or less) the same as products
> 
> You might be mixing up `/includes/classes/products.php` (which is deprecated) vs `/includes/classes/Product.php` (which replaces `products.php`). I think you've mostly cloned the new `Product.php` class, right? In that case, no deprecation issue.

Ah, upper case P vs lower case p, got it. Yes, it's a clone of Products so it's all good.



    CATEGORY_IMAGE_NO_IMAGE_STATUS
    CATEGORY_IMAGE_NO_IMAGE

I suppose these could be expanded to be used not only here with Featured Categories, but also in all the places where category images are used. Granted, it could be complicated to consistently determine whether it is a category vs a product image being affected, since the module that does it is kinda generic.
It would need a bit of a code study ... which my brain isn't ready to look into tonight yet.

They're almost all existing featured (there are a couple from brands but I won't tell if you won't) that have been renamed.
Tell you what, I'll put it all back and change the lang file to say featured products and categories.

lat9 commented on includes/modules/featured_categories.php:
```diff
@@ -63,7 +63,7 @@
     }
 
-    if (!$featured_categories->EOF) {
+    if ($featured_categories->EOF) {
```
Right, having traversed the items retrieved from the database, the returned database object will now be at `EOF`.  Given that this code section is already within a clause that's not going to be run if there aren't any featured categories, I'll suggest just removing the check on line 66 (and its closing brace) and sliding the code presently in that section to align with the ending brace for the `while` loop above.

lat9 commented on admin/featured_categories.php:
```diff
@@ -25,7 +25,7 @@
             if (isset($_POST['flag']) && ($_POST['flag'] === '1' || $_POST['flag'] === '0')) {
                 zen_set_featured_status((int)$_POST['id'], (int)$_POST['flag']);
-                zen_redirect(zen_href_link(FILENAME_FEATURED_CATEGORIES, zen_get_all_get_params(['action', 'fID']) . 'fID=' . $_POST['id'], 'NONSSL'));
+                zen_redirect(zen_href_link(FILENAME_FEATURED_CATEGORIES , zen_get_all_get_params(['action' , 'fID']) . 'fID=' . $_POST['id'] , 'NONSSL'));
```
Here (and a couple of other places 'downstream' in the code) you've added an unwanted space _before_ a function parameter's separating comma.  Line 28 was just fine (and proper) the way it was; please revert these.

There should be a space _after_ each parameter-separating comma, not before.

lat9 reviewed (changes requested):
Fix the leading commas in function parameters introduced in this change.

TMCSherpa commented on includes/templates/responsive_classic/sideboxes/tpl_categories.php:
```diff
@@ -60,7 +60,7 @@
 }
 
-if (SHOW_CATEGORIES_BOX_SPECIALS === 'true' || SHOW_CATEGORIES_BOX_PRODUCTS_NEW === 'true' || SHOW_CATEGORIES_BOX_FEATURED_PRODUCTS === 'true' || SHOW_CATEGORIES_BOX_PRODUCTS_ALL === 'true' || SHOW_CATEGORIES_BOX_FEATURED_CATEGORIES === 'true') {
+if (SHOW_CATEGORIES_BOX_SPECIALS == 'true' || SHOW_CATEGORIES_BOX_PRODUCTS_NEW == 'true' || SHOW_CATEGORIES_BOX_FEATURED_PRODUCTS == 'true' || SHOW_CATEGORIES_BOX_PRODUCTS_ALL == 'true' || SHOW_CATEGORIES_BOX_FEATURED_CATEGORIES == 'true') {
```
Oops, I reverted a bunch of code to figure out where the problem was (if ($show_this->RecordCount() > 0) to if (!$show_this->EOF), forgot the !) and forgot to put === back

lat9 reviewed (approved):
Thanks for those updates, @TMCSherpa

drbyte commented on includes/templates/template_default/templates/tpl_index_default.php:
```diff
@@ -32,7 +32,11 @@
-
+<?php
+/**
+ * display the Featured Categories Center Box
```
```suggestion
 * display the Featured Products Center Box
```

drbyte commented on includes/templates/template_default/templates/tpl_index_default.php:
```diff
@@ -32,7 +32,11 @@
-
+<?php
+/**
+ * display the Featured Categories Center Box
```
```suggestion
 * display the Featured Products Center Box
```

TMCSherpa replied:
main_page=featured_categories uses the display code from brands. I don't remember why but I'm sure it felt like a good idea at the time. Give me a few and I can replace it with the front page code.

drbyte commented on includes/templates/template_default/templates/tpl_index_default.php:
```diff
@@ -68,9 +72,19 @@
+<?php if ($show_display_category->fields['configuration_key'] == 'SHOW_PRODUCT_INFO_MAIN_FEATURED_CATEGORIES') { ?>
+<?php
+/**
+ * display the Featured Products Center Box
```
```suggestion
 * display the Featured Categories Center Box
```

TMCSherpa replied:
comments have been swapped

drbyte reviewed (changes requested):
A couple comments got set up backwards. Easy fixes.

Still exploring: the grid of featured categories on the main_page=featured_categories is a little compacted, not spaced out correctly. It might just be a CSS thing, or class not applied in the right spot.

drbyte commented:
@TMCSherpa This looks good. Thanks for all your work on it. 
Do you have any hesitations about merging? Anything else you think needs addressing?

TMCSherpa commented:
I think the only naughty thing left is I put featured catalog under featured products in the admin category menu rather than at the bottom so they both have the same sort order (13) if you update from 2.01. It is 14 and I renumbered everything afterwards if you do a fresh install

drbyte commented:
> I think the only naughty thing left is I put featured catalog under featured products in the admin category menu rather than at the bottom so they both have the same sort order (13) if you update from 2.01. It is 14 and I renumbered everything afterwards if you do a fresh install

Ya, I think it's fine like that.
